domain/secretbackend/state/state.go

```diff
@@ -0,0 +1,562 @@
+    $SecretBackendConfig.name,
+    $SecretBackendConfig.content
+)`, SecretBackendConfig{})
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		sb := SecretBackend{
+			UUID:        backend.ID,
+			Name:        backend.Name,
+			BackendType: backend.BackendType,
+		}
+		if backend.TokenRotateInterval != nil {
+			sb.TokenRotateInterval = domain.NullableDuration{Duration: *backend.TokenRotateInterval, Valid: true}
+		}
+		err := tx.Query(ctx, backendStmt, sb).Run()
+		if err != nil {
+			return fmt.Errorf("cannot insert secret backend %q: %w", backend.Name, err)
+		}
+		if backend.NextRotateTime != nil && !backend.NextRotateTime.IsZero() {
+			err = tx.Query(ctx, rotateStmt,
+				SecretBackendRotation{
+					BackendUUID:      backend.ID,
+					NextRotationTime: sql.NullTime{Time: *backend.NextRotateTime, Valid: true},
+				},
+			).Run()
+			if err != nil {
+				return fmt.Errorf("cannot insert secret backend rotation for %q: %w", backend.Name, err)
+			}
+		}
+		for k, v := range backend.Config {
+			err = tx.Query(ctx, configStmt,
+				SecretBackendConfig{
+					BackendUUID: backend.ID,
+					Name:        k,
+					Content:     v.(string),
+				},
+			).Run()
+			if err != nil {
+				return handleSecretBackendConfigError(err, backend.Name)
+			}
+		}
+		return err
+	})
+	return backend.ID, domain.CoerceError(err)
+}
+
+func handleSecretBackendConfigError(err error, identifier string) error {
+	if err == nil {
+		return nil
+	}
+	if database.IsErrConstraintCheck(err) {
+		return fmt.Errorf(
+			"cannot upsert secret backend config for %q: %w",
+			identifier, errors.NewNotValid(nil, "empty config name or content"),
+		)
+	}
+	return fmt.Errorf("cannot upsert secret backend config for %q: %w", identifier, err)
+}
+
+// UpdateSecretBackend updates an existing secret backend.
+func (s *State) UpdateSecretBackend(ctx context.Context, backend secretbackend.UpdateSecretBackendParams) error {
+	if backend.ID == "" {
+		return errors.NewNotValid(nil, "backend ID is missing")
+	}
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	nameStmt, err := sqlair.Prepare(`
+UPDATE secret_backend
+SET name = $SecretBackend.name
+WHERE uuid = $SecretBackend.uuid`, SecretBackend{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	rotateIntervalStmt, err := sqlair.Prepare(`
+UPDATE secret_backend
+SET token_rotate_interval = $SecretBackend.token_rotate_interval
+WHERE uuid = $SecretBackend.uuid`, SecretBackend{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	rotationTimeStmt, err := sqlair.Prepare(`
+UPDATE secret_backend_rotation
+SET next_rotation_time = $SecretBackendRotation.next_rotation_time
+WHERE backend_uuid = $SecretBackendRotation.backend_uuid`, SecretBackendRotation{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	configStmt, err := sqlair.Prepare(`
+INSERT INTO secret_backend_config (backend_uuid, name, content) VALUES
```

Comment:
We need to delete existing config that doesn't exist in the new config - see storage pool attrs etc